helix-term/src/ui/picker.rs

```diff
@@ -532,27 +536,27 @@ impl<T: 'static + Send + Sync, D: 'static + Send + Sync> Picker<T, D> {
         }
     }
 
-    fn current_file(&self, editor: &Editor) -> Option<FileLocation> {
-        self.selection()
-            .and_then(|current| (self.file_fn.as_ref()?)(editor, current))
-            .map(|(path_or_id, line)| (path_or_id.get_canonicalized(), line))
-    }
-
-    /// Get (cached) preview for a given path. If a document corresponding
+    /// Get (cached) preview for the currently selected item. If a document corresponding
     /// to the path is already open in the editor, it is used instead.
     fn get_preview<'picker, 'editor>(
         &'picker mut self,
-        path_or_id: PathOrId,
         editor: &'editor Editor,
-    ) -> Preview<'picker, 'editor> {
+    ) -> Option<(Preview<'picker, 'editor>, Option<(usize, usize)>)> {
+        let current = self.selection()?;
+        let (path_or_id, range) = (self.file_fn.as_ref()?)(editor, current)?;
+
         match path_or_id {
             PathOrId::Path(path) => {
-                if let Some(doc) = editor.document_by_path(&path) {
-                    return Preview::EditorDocument(doc);
+                let path = path.as_ref();
+                if let Some(doc) = editor.document_by_path(path) {
+                    return Some((Preview::EditorDocument(doc), range));
                 }
 
-                if self.preview_cache.contains_key(&path) {
-                    let preview = &self.preview_cache[&path];
+                if self.preview_cache.contains_key(path) {
```

Comment:
hmm I actually implemented this locally IIRC.

Reply:
``` rust
                if let Some((path, preview)) = self.preview_cache.get_key_value(path) {
                    if matches!(preview, CachedPreview::Document(doc) if doc.language_config().is_none())
                    {
                        helix_event::send_blocking(&self.preview_highlight_handler, path.clone());
                    }
                    return Preview::Cached(preview);
                }

```

this is what I have here, maybe myBranch didn't get far enough to get compilation errors maybe this helps (another nit I had here was to use matches when there is only a single branch being used)

Reply:
For this I see:

```text
error[E0502]: cannot borrow `self.preview_cache` as mutable because it is also borrowed as immutable
   --> helix-term/src/ui/picker.rs:595:17
    |
541 |     fn get_preview<'picker, 'editor>(
    |                    ------- lifetime `'picker` defined here
...
555 |                 if let Some((path, preview)) = self.preview_cache.get_key_value(path) {
    |                                                -------------------------------------- immutable borrow occurs here
...
563 |                     return Some((Preview::Cached(preview), range));
    |                            --------------------------------------- returning this value requires that `self.preview_cache` is borrowed for `'picker`
...
595 |                 self.preview_cache.insert(path.clone(), preview);
    |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ mutable borrow occurs here

For more information about this error, try `rustc --explain E0502`.
```

I think this error might not pop up on your branch since these lifetime checks seem to come fairly late in the compilation process (so if there are earlier errors, the compiler doesn't seem to emit them).

The `matches!` part though is doable and a nice readability improvement :+1:

Reply:
yeah that makes sense, looking forward to the day the borrowchecker can handle these cases. 

You may be  able to massage code like to work (I think I had some weird hacks that fixed stuff like this in a different case but at that point that's probably not better than the current code.